metals/src/main/scala/scala/meta/internal/metals/codeactions/InlineValueCodeAction.scala

```diff
@@ -0,0 +1,180 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Defn
+import scala.meta.Pat
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta.internal.metals.Compilers
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.ServerCommands
+import scala.meta.internal.metals.clients.language.MetalsLanguageClient
+import scala.meta.internal.metals.codeactions.CodeAction
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.{lsp4j => l}
+
+class InlineValueCodeAction(
+    trees: Trees,
+    compilers: Compilers,
+    languageClient: MetalsLanguageClient,
+) extends CodeAction {
+
+  type CommandData = ServerCommands.InlineValueParams
+
+  override def command: Option[ActionCommand] =
+    Some(ServerCommands.InlineValue)
+
+  override def kind: String = l.CodeActionKind.RefactorInline
+
+  override def contribute(params: l.CodeActionParams, token: CancelToken)(
+      implicit ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val pathStr = params.getTextDocument.getUri
+    val path = pathStr.toAbsolutePath
+    val range = params.getRange()
+    val action =
+      for {
+        termName <- getTermNameForPos(path, range)
+        if (hasDefinitionInScope(termName))
+        optThisDef = getDefinitionOfValue(path, range)
+        if (optThisDef.map(isLocal(_)).getOrElse(true))
+      } yield {
+        val command =
+          ServerCommands.InlineValue.toLsp(
+            ServerCommands.InlineValueParams(
+              params.getTextDocument(),
+              termName.pos.toLsp,
+              optThisDef.isDefined,
+            )
+          )
+        CodeActionBuilder.build(
+          title = InlineValueCodeAction.title(termName.value),
+          kind = this.kind,
+          command = Some(command),
+        )
+      }
+    action.toSeq
+  }
+
+  private def isLocal(definition: Defn.Val): Boolean =
+    definition.parent match {
+      case Some(_: Term.Block) => true
+      case _ => false
+    }
+
+  // Check if the definition is in the scope (in the same file)
+  // NOTE: 1. import from e.g. `Object` in the file will return false
+  //       2. can have FALSE POSITIVES but the condition is checked properly upon execution
+  private def hasDefinitionInScope(
+      nameTerm: Term.Name
+  ): Boolean = {
+    def isDefinition(t: Tree): Boolean = {
+      t match {
+        case v: Defn.Val =>
+          v.pats.exists {
+            case p: Pat.Var if (p.name.value == nameTerm.value) => true
```

Comment:
I feel like we wouldn't want to inline such case anyway, 'cause we would have to transform this pattern match extraction into an "inline" one. E.g.
```
val Some(i) = optCase
val g = i
```
would have to be
```
val g = optCase.get()
```
That sounds like something beyond the scope of what we are doing here.

Reply:
You're right, so maybe we should do the opposite. Add additional condition, something like:
```
case Defn.Val(_, List(Pat.Var(name)), _, _) =>
          name.value == nameTerm.value
```
so that we accept only a single val. Or handle things like `val a, b = 123`

Reply:
True, I don't think I realised that this is why pats are a list. I kind of want to go for the single use, since `val a, b = 123` feels like a code that no one should really have a use for.